### ISSUE TITLE
Call TestNGCucumberTests.finish() throw java.io.IOException: Stream closed when using TestNG invocation

### ISSUE THREAD

**Issue by thanhlelgg** [cucumber/cucumber-jvm#2080]
**Describe the bug**
I'm trying to run a test multiple time with invocation from TestNG then generate a json report. But if `invocationCount` is more than 1, then TestNGCucumberTests.finish() throw out `java.lang.RuntimeException: java.io.IOException: Stream closed`

**To Reproduce**

Test Runner: 
```java
public abstract class TestNGCucumberTests {
    private TestNGCucumberRunner testNGCucumberRunner;
    private ScenarioContext scenarioContext = new ScenarioContext();

    @BeforeClass(alwaysRun = true)
    public void setUpClass() {
        testNGCucumberRunner = new TestNGCucumberRunner(this.getClass());
    }

    @Test(groups = "cucumber", description = "Runs Cucumber Scenarios", dataProvider = "scenarios", alwaysRun = true, invocationCount=2)
    public void runScenario(PickleEventWrapper pickleWrapper, CucumberFeatureWrapper featureWrapper) throws Throwable {
        scenarioContext.processScenario(pickleWrapper.getPickleEvent());
        testNGCucumberRunner.runScenario(pickleWrapper.getPickleEvent());
    }

    /**
     * Returns two dimensional array of PickleEventWrapper scenarios with their
     * associated CucumberFeatureWrapper feature.
     *
     * @return a two dimensional array of scenarios features.
     */
    @DataProvider()
    public Object[][] scenarios() {
        if (testNGCucumberRunner == null) {
            return new Object[0][0];
        }
        return testNGCucumberRunner.provideScenarios();
    }

    @AfterClass(alwaysRun = true)
    public void tearDownClass() {
        if (testNGCucumberRunner == null) {
            return;
        }
        testNGCucumberRunner.finish();
    }
}
```

Test Class:
```java
@CucumberOptions(features = {"src/test/resources/features/...feature"},
        glue = {"com/..."}, plugin = {"json:test-output/json/path_to_file.json"})
public class CucumberTests extends TestNGCucumberTests {
}
```

Test Suite xml:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="Suite" thread-count="1" parallel="classes">
    <listeners>
        <listener class-name="com.directasia.utils.TestListener"/>
    </listeners>
    <test name="Testing">
        <parameter name="browser" value="chrome.local"/>
        <classes>
            <class name="com.path.to.tests"/>
        </classes>
    </test>
</suite>
```

Run the xml file will print out this error:
```
java.lang.RuntimeException: java.io.IOException: Stream closed

	at cucumber.api.formatter.NiceAppendable.append(NiceAppendable.java:34)
	at cucumber.api.formatter.NiceAppendable.append(NiceAppendable.java:10)
	at gherkin.deps.com.google.gson.internal.Streams$AppendableWriter.write(Streams.java:90)
	at java.io.Writer.write(Writer.java:192)
	at java.io.Writer.write(Writer.java:157)
	at gherkin.deps.com.google.gson.stream.JsonWriter.open(JsonWriter.java:327)
	at gherkin.deps.com.google.gson.stream.JsonWriter.beginArray(JsonWriter.java:288)
	at gherkin.deps.com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.write(CollectionTypeAdapterFactory.java:94)
	at gherkin.deps.com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.write(CollectionTypeAdapterFactory.java:60)
	at gherkin.deps.com.google.gson.Gson.toJson(Gson.java:600)
	at gherkin.deps.com.google.gson.Gson.toJson(Gson.java:579)
	at gherkin.deps.com.google.gson.Gson.toJson(Gson.java:554)
	at cucumber.runtime.formatter.JSONFormatter.finishReport(JSONFormatter.java:173)
	at cucumber.runtime.formatter.JSONFormatter.access$400(JSONFormatter.java:42)
	at cucumber.runtime.formatter.JSONFormatter$5.receive(JSONFormatter.java:82)
	at cucumber.runtime.formatter.JSONFormatter$5.receive(JSONFormatter.java:79)
	at cucumber.runner.AbstractEventPublisher.send(AbstractEventPublisher.java:45)
	at cucumber.runner.AbstractEventPublisher.sendAll(AbstractEventPublisher.java:52)
	at cucumber.runner.CanonicalOrderEventPublisher.handle(CanonicalOrderEventPublisher.java:18)
	at cucumber.runtime.formatter.Plugins$1.receive(Plugins.java:53)
	at cucumber.runner.AbstractEventPublisher.send(AbstractEventPublisher.java:38)
	at cucumber.runner.AbstractEventBus.send(AbstractEventBus.java:9)
	at cucumber.runner.TimeServiceEventBus.send(TimeServiceEventBus.java:3)
	at io.cucumber.testng.TestNGCucumberRunner.finish(TestNGCucumberRunner.java:88)
	at com.directasia.utils.TestNGCucumberTests.tearDownClass(TestNGCucumberTests.java:48)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:134)
	at org.testng.internal.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:63)
	at org.testng.internal.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:348)
	at org.testng.internal.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:302)
	at org.testng.internal.TestMethodWorker.invokeAfterClassMethods(TestMethodWorker.java:217)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:130)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Stream closed
	at sun.nio.cs.StreamEncoder.ensureOpen(StreamEncoder.java:45)
	at sun.nio.cs.StreamEncoder.write(StreamEncoder.java:118)
	at sun.nio.cs.StreamEncoder.write(StreamEncoder.java:135)
	at java.io.OutputStreamWriter.write(OutputStreamWriter.java:220)
	at java.io.Writer.write(Writer.java:157)
	at java.io.Writer.append(Writer.java:269)
	at java.io.Writer.append(Writer.java:50)
	at cucumber.api.formatter.NiceAppendable.append(NiceAppendable.java:30)
	... 37 more
```

**Expected behavior**
Json report file should be exported correctly

**Environment**
Windows 10 Pro
Cucumber version 4.8.1

**Note**
Run test in parallel or single thread both throw the error.
If I remove the json report plugin or don't use the invocation, then there will be no problem.

**Comment by mpkorstanje** [comment]
Could you try to reproduce this with the latest version of Cucumber?

**Comment by mpkorstanje** [comment]
Please feel free to re open with a minimal reproducer on a recent Cucumber version.

**Comment by lprimak** [comment]
I believe this is unrelated to cucumber, but an arquillian issue. It maybe solved with 1.7 Alpha release but I am not quite sure about that
